proxylib/proxylib/policymap.go

```diff
@@ -0,0 +1,332 @@
+	// to pass through since the bpf datapath already allowed it.
+	// TODO: Change back to false only when non-bpf datapath is supported?
+
+	//	log.Debugf("NPDS::PortNetworkPolicies(port=%d, remoteId=%d): allowing traffic on port for which there is no policy, assuming L3/L4 has passed it! (%v)", port, remoteId, p)
+	//	return !(found || foundWc)
+	if !(found || foundWc) {
+		log.Debugf("NPDS::PortNetworkPolicies(port=%d, remoteId=%d): Dropping traffic on port for which there is no policy! (%v)", port, remoteId, p)
+	}
+	return false
+}
+
+type PolicyInstance struct {
+	protobuf cilium.NetworkPolicy
+	Ingress  PortNetworkPolicies
+	Egress   PortNetworkPolicies
+}
+
+func newPolicyInstance(config *cilium.NetworkPolicy) *PolicyInstance {
+	log.Debugf("NPDS::PolicyInstance: Inserting policy %s", config.String())
+	
+	return &PolicyInstance{
+		protobuf: *config,
+		Ingress:  newPortNetworkPolicies(config.GetIngressPerPortPolicies()),
+		Egress:   newPortNetworkPolicies(config.GetEgressPerPortPolicies()),
+	}
+}
+
+func (p *PolicyInstance) Matches(ingress bool, port, remoteId uint32, l7 interface{}) bool {
+	if ingress {
+		return p.Ingress.Matches(port, remoteId, l7)
+	}
+	return p.Egress.Matches(port, remoteId, l7)
+}
+
+type PolicyMap map[string]*PolicyInstance
+
+var policyMap unsafe.Pointer // PolicyMap
+
+func newPolicyMap() PolicyMap {
+	return make(PolicyMap)
+}
+
+func getPolicyMap() *PolicyMap {
+	return (*PolicyMap)(atomic.LoadPointer(&policyMap))
+}
+
+func PolicyMatches(endpointPolicyName string, ingress bool, port, remoteId uint32, l7 interface{}) bool {
+	// Policy maps are never modified once published
+	pMap := getPolicyMap()
+	if pMap == nil {
+		log.Warning("NPDS: No Policy Received!")
+		return false
+	}
+	policy, found := (*pMap)[endpointPolicyName]
+
+	if !found {
+		log.Warningf("NPDS: Policy for %s not found (%v)", endpointPolicyName, *pMap)
+	}
+
+	return found && policy.Matches(ingress, port, remoteId, l7)
+}
+
+// Update the PolicyMap from a protobuf. PolicyMap is only ever changed if the whole update is successful.
+// Assuming there is only one updater!
+func PolicyUpdate(resp *envoy_api_v2.DiscoveryResponse) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			var ok bool
+			if err, ok = r.(error); !ok {
+				err = fmt.Errorf("NPDS: Panic: %v", r)
+			}
+		}
+	}()
+
+	log.Debugf("NPDS: Updating policy from %v", resp)
+
+	oldMap := getPolicyMap()
+	newMap := newPolicyMap()
+
+	for _, any := range resp.Resources {
+		if any.TypeUrl != resp.TypeUrl {
+			return fmt.Errorf("NPDS: Mismatching TypeUrls: %s != %s", any.TypeUrl, resp.TypeUrl)
+		}
+		var config cilium.NetworkPolicy
+		if err = proto.Unmarshal(any.Value, &config); err != nil {
+			return fmt.Errorf("NPDS: Policy unmarshal error: %v", err)
+		}
+
+		policyName := config.GetName()
+
+		if oldMap != nil {
+			// Locate the old version, if any
+			oldPolicy, found := (*oldMap)[policyName]
```

Comment:
[1] will avoid you doing things like these `(*oldMap)[policyName]`

Reply:
Also changed to init() the map so the old map is newer nil. Simplifies a bit.